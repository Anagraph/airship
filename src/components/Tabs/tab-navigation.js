import React from 'react';

const TabNavigation = ({ labels, selected, clickHandler }) => {
  const links = labels.map((label, index) => {
    const css = index === selected ? 'active' : '';
    return (
      <li key={label}>
        <button className={css} onClick={e => clickHandler(e, label)}>
          {label}
        </button>
      </li>
    );
  });
  return <ul>{links}</ul>;
};

export default TabNavigation;
