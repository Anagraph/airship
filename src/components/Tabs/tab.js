import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { colors } from '../../constants';
/*
  <Tabs>
    <Tabs.Panel label="wadus">
      Hola
    </Tabs.Panel>
    <Tabs.Panel label="world">
      Mundo
    </Tabs.Panel>
  </Tabs>
*/

const font = props => {
  let font = "400 12px/20px 'Roboto'";
  if (props.large) {
    font = "400 16px/24px 'Roboto'";
  }
  return font;
};

const shadow = props => (props.large ? '-4px' : '-2px');

const StyledList = styled.ul`
  list-style: none;
  display: flex;
  margin: 0;
  padding: 0;
  box-shadow: inset 0 -1px 0 0 ${colors.grey};
`;

const StyledButton = styled.button`
  border: 0;
  background-color: transparent;
  color: ${colors.primaryColor};
  font: ${font};
  padding: 8px;

  &:focus {
    outline: none;
  }

  &:hover {
    box-shadow: inset 0 ${shadow} 0 0 ${colors.complementaryColor};
  }

  &.is-selected,
  &.is-selected:hover {
    box-shadow: inset 0 ${shadow} 0 0 ${colors.black};
    color: ${colors.black};
  }
`;
StyledButton.displayName = 'StyledButton';

const Tabpanel = ({ children, selected }) => (selected ? <div>{children}</div> : null);

Tabpanel.propTypes = {
  children: PropTypes.node,
  label: PropTypes.string.isRequired,
  selected: PropTypes.bool,
};
Tabpanel.displayName = 'Tabpanel';

class Tabs extends Component {
  static defaultProps = {
    selected: 0,
  };
  initialState = {
    selected: this.props.selected,
  };
  state = this.initialState;

  static Panel = Tabpanel;

  constructor(props) {
    super(props);
    this.navigation = this.buildNavigation();
  }

  componentDidMount() {
    const { selected } = this.props;

    if (typeof selected === 'string') {
      this.setSelected(selected);
    }
  }

  buildNavigation() {
    const { children } = this.props;
    return React.Children.map(children, child => child.props.label);
  }

  setSelected = label => {
    const index = this.navigation.indexOf(label);
    const { onChange } = this.props;
    this.setState(
      state => ({ ...state, selected: index }),
      () => onChange && onChange(this.state),
    );
  };

  render() {
    const { selected } = this.state;
    const { children, large } = this.props;

    return (
      <div>
        <StyledList>
          {this.navigation.map((label, index) => {
            const css = selected === index ? 'is-selected' : '';
            return (
              <li key={label}>
                <StyledButton
                  className={css}
                  onClick={() => this.setSelected(label)}
                  large={large}
                >
                  {label}
                </StyledButton>
              </li>
            );
          })}
        </StyledList>
        {React.Children.map(children, (child, i) => React.cloneElement(child, {
            selected: i === selected,
          }))}
      </div>
    );
  }
}

Tabs.propTypes = {
  children: PropTypes.node,
  large: PropTypes.bool,
  selected: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func,
};

export default Tabs;
