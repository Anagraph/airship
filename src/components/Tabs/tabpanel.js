import React from 'react';

const Tabpanel = ({ children, selected }) => {
  const css = selected ? 'active' : '';
  return <div className={css}>{children}</div>;
};

export default Tabpanel;
