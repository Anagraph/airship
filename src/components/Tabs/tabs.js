import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { theme } from '../../constants';

const font = props => {
  let font = "400 12px/20px 'Roboto'";
  if (props.large) {
    font = "400 16px/24px 'Roboto'";
  }
  return font;
};

const shadow = props => (props.large ? '-4px' : '-2px');

const TabList = styled.ul`
  list-style: none;
  display: flex;
  margin: 0;
  padding: 0;
  box-shadow: inset 0 -1px 0 0 ${props => props.theme.ui03};
`;
TabList.defaultProps = {
  theme,
};

const TabButton = styled.button`
  border: 0;
  background-color: transparent;
  color: ${props => props.theme.brand01};
  font: ${font};
  padding: 8px;
  cursor: pointer;
  margin: 0;

  &:focus {
    outline: none;
  }

  &:hover {
    box-shadow: inset 0 ${shadow} 0 0 ${props => props.theme.brand03};
  }

  ${props => props.selected && css`
    &,
    &:hover {
      box-shadow: inset 0 ${shadow} 0 0 ${props => props.theme.type01};
      color: ${props => props.theme.type01};
    }
  `};
`;
TabButton.displayName = 'TabButton';
TabButton.defaultProps = {
  theme,
};

const Tabpanel = ({ children, selected }) => (selected ? <div>{children}</div> : null);

Tabpanel.propTypes = {
  children: PropTypes.node,
  label: PropTypes.string.isRequired,
  selected: PropTypes.bool,
};
Tabpanel.displayName = 'Tabpanel';

class Tabs extends Component {
  static Panel = Tabpanel;

  static defaultProps = {
    selected: 0,
  };

  initialState = {
    selected: this.props.selected,
  };

  state = this.initialState;

  componentDidMount() {
    const { selected } = this.props;

    if (typeof selected === 'string') {
      this.setSelected(selected);
    }
  }

  getNavigation() {
    return React.Children.map(this.props.children, child => child.props.label);
  }

  setSelected = label => {
    const index = this.getNavigation().indexOf(label);
    const { onChange } = this.props;

    if (index === this.state.selected) return;

    this.setState(
      state => ({ ...state, selected: index }),
      () => onChange && onChange(this.state),
    );
  };

  render() {
    const { selected } = this.state;
    const { children, large } = this.props;

    return (
      <div>
        <TabList>
          {this.getNavigation().map((label, index) => (
            <li key={label}>
              <TabButton
                onClick={() => this.setSelected(label)}
                large={large}
                selected={selected === index}
              >
                {label}
              </TabButton>
            </li>
          ))}
        </TabList>

        {React.Children.map(children, (child, i) => React.cloneElement(child, {
            selected: i === selected,
          }))}
      </div>
    );
  }
}

Tabs.propTypes = {
  children: PropTypes.node,
  large: PropTypes.bool,
  selected: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func,
};

export default Tabs;
