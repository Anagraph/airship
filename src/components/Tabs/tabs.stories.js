import React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import Tabs from './tabs';
import Text from '../Typography/text';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  brand01: '#FABADA',
  brand03: '#FABADA',
};

storiesOf('Tabs', module)
  .add('Default', () => (
    <Tabs>
      <Tabs.Panel label="First tab">
        <Text>Hello there</Text>
      </Tabs.Panel>
      <Tabs.Panel label="Second tab">
        <Text>General Kenobi</Text>
      </Tabs.Panel>
    </Tabs>
  ))
  .add('With custom theme', () => (
    <ThemeProvider theme={CUSTOM_THEME}>
      <Tabs>
        <Tabs.Panel label="First tab">
          <Text>Hello there</Text>
        </Tabs.Panel>
        <Tabs.Panel label="Second tab">
          <Text>General Kenobi</Text>
        </Tabs.Panel>
      </Tabs>
    </ThemeProvider>
  ));
