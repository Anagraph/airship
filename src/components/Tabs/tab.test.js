import React from 'react';
import Tabs from './tab';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

describe('render', () => {
  it('renders without crashing', () => {
    const component = renderer.create(
      <Tabs>
        <Tabs.Panel label="wadus">Hola</Tabs.Panel>
        <Tabs.Panel label="world">Mundo</Tabs.Panel>
      </Tabs>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with large param', () => {
    const component = renderer.create(
      <Tabs large>
        <Tabs.Panel label="wadus">Hola</Tabs.Panel>
        <Tabs.Panel label="world">Mundo</Tabs.Panel>
      </Tabs>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with selected as number', () => {
    const component = renderer.create(
      <Tabs selected={1}>
        <Tabs.Panel label="wadus">Hola</Tabs.Panel>
        <Tabs.Panel label="world">Mundo</Tabs.Panel>
      </Tabs>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with selected as string', () => {
    const component = renderer.create(
      <Tabs selected="world">
        <Tabs.Panel label="wadus">Hola</Tabs.Panel>
        <Tabs.Panel label="world">Mundo</Tabs.Panel>
      </Tabs>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call callback when changed', () => {
    const spy = jest.fn();
    const component = shallow(
      <Tabs selected="world" onChange={spy}>
        <Tabs.Panel label="wadus">Hola</Tabs.Panel>
        <Tabs.Panel label="world">Mundo</Tabs.Panel>
      </Tabs>
    );

    const button = component.find('StyledButton').first();
    button.simulate('click');
    expect(spy).toHaveBeenCalled();
  });
});
