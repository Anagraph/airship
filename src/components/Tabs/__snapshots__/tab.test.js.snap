// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders with large param 1`] = `
<div>
  <ul
    className="sc-bdVaJa eLvrGR"
  >
    <li>
      <button
        className="is-selected sc-bwzfXH jEjPnu"
        onClick={[Function]}
      >
        wadus
      </button>
    </li>
    <li>
      <button
        className="sc-bwzfXH jEjPnu"
        onClick={[Function]}
      >
        world
      </button>
    </li>
  </ul>
  <div>
    Hola
  </div>
</div>
`;

exports[`render renders with selected as number 1`] = `
<div>
  <ul
    className="sc-bdVaJa eLvrGR"
  >
    <li>
      <button
        className="sc-bwzfXH gOikWb"
        onClick={[Function]}
      >
        wadus
      </button>
    </li>
    <li>
      <button
        className="is-selected sc-bwzfXH gOikWb"
        onClick={[Function]}
      >
        world
      </button>
    </li>
  </ul>
  <div>
    Mundo
  </div>
</div>
`;

exports[`render renders with selected as string 1`] = `
<div>
  <ul
    className="sc-bdVaJa eLvrGR"
  >
    <li>
      <button
        className="sc-bwzfXH gOikWb"
        onClick={[Function]}
      >
        wadus
      </button>
    </li>
    <li>
      <button
        className="is-selected sc-bwzfXH gOikWb"
        onClick={[Function]}
      >
        world
      </button>
    </li>
  </ul>
  <div>
    Mundo
  </div>
</div>
`;

exports[`render renders without crashing 1`] = `
<div>
  <ul
    className="sc-bdVaJa eLvrGR"
  >
    <li>
      <button
        className="is-selected sc-bwzfXH gOikWb"
        onClick={[Function]}
      >
        wadus
      </button>
    </li>
    <li>
      <button
        className="sc-bwzfXH gOikWb"
        onClick={[Function]}
      >
        world
      </button>
    </li>
  </ul>
  <div>
    Hola
  </div>
</div>
`;
